MENTAL-MODEL.md

```diff
@@ -0,0 +1,376 @@
+
+The `danger_accept_invalid_certs` parameter is used for testing purposes only detailed in sectino 5.
+
+### 1. Parse BIP21 as `payjoin::Uri`
+
+Start by parsing a valid BIP 21 uri having the `pj` parameter. This is the [`bip21`](https://crates.io/crates/bip21) crate under the hood.
+
+```rs
+let link = payjoin::Uri::try_from(bip21)
+    .map_err(|e| anyhow!("Failed to create URI from BIP21: {}", e))?;
+
+let link = link
+    .check_pj_supported()
+    .map_err(|e| anyhow!("The provided URI doesn't support payjoin (BIP78): {}", e))?;
+```
+
+### 2. Construct URI request parameters, a finalized "Original PSBT" paying `.amount` to `.address`
+
+```rs
+let mut outputs = HashMap::with_capacity(1);
+outputs.insert(link.address.to_string(), amount);
+
+let options = bitcoincore_rpc::json::WalletCreateFundedPsbtOptions {
+    lock_unspent: Some(true),
+    fee_rate: Some(Amount::from_sat(2000)), // SHOULD CHANGE TO AN OPTIONAL FEE RATE
+    ..Default::default()
+};
+let psbt = bitcoind
+    .wallet_create_funded_psbt(
+        &[], // inputs
+        &outputs,
+        None, // locktime
+        Some(options),
+        None,
+    )
+    .context("Failed to create PSBT")?
+    .psbt;
+let psbt = bitcoind
+    .wallet_process_psbt(&psbt, None, None, None)
+    .with_context(|| "Failed to process PSBT")?
+    .psbt;
+let psbt = load_psbt_from_base64(psbt.as_bytes()) // SHOULD BE PROVIDED BY CRATE AS HELPER USING rust-bitcoin base64 feature
+    .with_context(|| "Failed to load PSBT from base64")?;
+log::debug!("Original psbt: {:#?}", psbt);
+let pj_params = payjoin::sender::Configuration::with_fee_contribution(
+    payjoin::bitcoin::Amount::from_sat(10000),
+    None,
+);
+```
+
+### 3. (optional) Spawn a thread or async task that will broadcast the transaction after one minute unless canceled
+
+I wrote this in the original docs, but I think it should be amended.
+
+In case the payjoin goes through but you still want to pay by default. This missing `payjoin-client`
+
+Writing this, I think of [Signal's contributing guidelines](https://github.com/signalapp/Signal-iOS/blob/main/CONTRIBUTING.md#development-ideology):
+> The answer is not more options. If you feel compelled to add a preference that's exposed to the user, it's very possible you've made a wrong turn somewhere.
+
+### 4. Construct the request with the PSBT and parameters
+
+```rs
+let (req, ctx) = link
+    .create_pj_request(psbt, pj_params)
+    .with_context(|| "Failed to create payjoin request")?;
+```
+
+### 5. Send the request and receive response
+
+Senders request a payjoin from the receiver with a payload containing the Original PSBT  and optional parameters. They require a secure endpoint for authentication and message secrecy to prevent that transaction from being modified by a malicious third party during transit or being snooped on. Only https and .onion endpoints are spec-compatible payjoin endpoints.
+
+Avoiding the secure endpoint requirement is convenient for testing.
```

Comment:
(but again, Signal guidelines remind me of the folly of options). I wonder if there's a better way to set up a testing environment with a valid https cert with something like Nix.